packages/miniflare/src/zod-format.ts

```diff
@@ -0,0 +1,296 @@
+
+    // Print actual value
+    const actual = util.inspect(annotated[kActual], inspectOptions);
+    const actualIndented = actual
+      .split("\n")
+      .map((line, i) => (i > 0 ? prefixIndent + line : line))
+      .join("\n");
+
+    // Print message
+    let messageColour = red;
+    let messagePrefix = prefixIndent + "^";
+    let groupOr = "";
+    if (annotated[kGroupId] !== undefined) {
+      // If this annotation was part of a group, set the message colour based
+      // on the group, and include the group ID in the prefix
+      messageColour = groupColours[annotated[kGroupId] % groupColours.length];
+      messagePrefix += annotated[kGroupId] + 1;
+      const remaining = groupCounts.get(annotated[kGroupId]);
+      assert(remaining !== undefined);
+      if (remaining > 1) groupOr = " *or*";
+      groupCounts.set(annotated[kGroupId], remaining - 1);
+    }
+    messagePrefix += " ";
+
+    const messageIndent = " ".repeat(messagePrefix.length);
+    const messageIndented = annotated[kMessages]
+      .flatMap((m) => m.split("\n"))
+      .map((line, i) => (i > 0 ? messageIndent + line : line))
+      .join("\n");
+
+    // Print final annotation
+    const error = messageColour(`${messagePrefix}${messageIndented}${groupOr}`);
+    return `${indent}${dim(prefix)}${actualIndented}${dim(suffix)}\n${error}`;
+  } else if (Array.isArray(annotated)) {
+    // Print array recursively
+    let result = `${indent}${dim(`${prefix}[`)}\n`;
+    const arrayIndent = indent + "  ";
+    for (let i = 0; i < annotated.length; i++) {
+      const value = annotated[i];
+      // Add `...` if the last item wasn't an `...`
+      if (value === undefined && (i === 0 || annotated[i - 1] !== undefined)) {
+        result += `${arrayIndent}${dim("...,")}\n`;
+      }
+      if (value !== undefined) {
+        result += print(inspectOptions, groupCounts, value, arrayIndent, {
+          prefix: `/* [${i}] */ `,
+          suffix: ",",
+        });
+        result += "\n";
+      }
+    }
+    result += `${indent}${dim(`]${suffix}`)}`;
+    return result;
+  } else {
+    // Print object recursively
+    assert(isRecord(annotated));
+    let result = `${indent}${dim(`${prefix}{`)}\n`;
+    const objectIndent = indent + "  ";
+    const entries = Object.entries(annotated);
+    for (let i = 0; i < entries.length; i++) {
+      const [key, value] = entries[i];
+      // Add `...` if the last item wasn't an `...`
+      if (value === undefined && (i === 0 || entries[i - 1][1] !== undefined)) {
+        result += `${objectIndent}${dim("...,")}\n`;
+      }
+      if (value !== undefined) {
+        result += print(inspectOptions, groupCounts, value, objectIndent, {
+          prefix: `${key}: `,
+          suffix: ",",
+        });
+        result += "\n";
+      }
+    }
+    result += `${indent}${dim(`}${suffix}`)}`;
+    return result;
+  }
+}
+
+export function formatZodError(error: z.ZodError, input: unknown): string {
```

Comment:
Did you consider using https://github.com/colinhacks/zod/blob/master/ERROR_HANDLING.md#formatting-errors, which gives you ZodErrors in the form: 

```ts
if (!result.success) {
  console.log(result.error.format());
  /*
    {
      name: {
        _errors: ['Expected string, received null']
      },
      contactInfo: {
        email: {
          _errors: ['Invalid email']
        }
      }
    }
  */
}
```